Quest crashes when it loads photo capture widget

**Describe the bug**
Quest app crashes when it tries to load a questionnaire from HAPI with photo capture widget in it.

**To Reproduce**
Steps to reproduce the behavior:
1. Go to Quest app
2. Login
2. Click on one of the clients
3. Click on 'G6PD TEST PHOTO RESULT'
4. Wait and then crash

**Expected behavior**
Quest app should not crash and successfully loads the questionnaire

**Screenshots**
<img src="https://user-images.githubusercontent.com/62053304/142377176-8e53547b-f6e1-4008-8a5f-472f254797ae.png" width="256"/>
<img src="https://user-images.githubusercontent.com/62053304/142377190-aafc14cf-b656-4963-8792-158a1eeccdab.png" width="256"/>
<img src="https://user-images.githubusercontent.com/62053304/142377165-ac2031bf-1a6f-48ba-863d-a5b98f170877.png" width="256"/>
<img src="https://user-images.githubusercontent.com/62053304/142377171-196e23f6-0ba4-4943-b916-2c564454ce53.png" width="256"/>

**Smartphone (please complete the following information):**
 - Device: Pixel 4 emulator
 - Android version: 11
 - Build number: RSR1.201013.001
 - Version API 30

**Additional context**
```
2021-11-18 14:53:21.390 21398-21398/org.smartregister.fhircore.quest E/AndroidRuntime: FATAL EXCEPTION: main
    Process: org.smartregister.fhircore.quest, PID: 21398
    java.lang.RuntimeException: An exception happened in constructor of class com.google.android.fhir.datacapture.QuestionnaireViewModel
        at androidx.lifecycle.SavedStateViewModelFactory.create(SavedStateViewModelFactory.java:132)
        at androidx.lifecycle.ViewModelProvider.get(ViewModelProvider.java:185)
        at androidx.lifecycle.ViewModelProvider.get(ViewModelProvider.java:150)
        at androidx.lifecycle.ViewModelLazy.getValue(ViewModelProvider.kt:54)
        at androidx.lifecycle.ViewModelLazy.getValue(ViewModelProvider.kt:41)
        at com.google.android.fhir.datacapture.QuestionnaireFragment.getViewModel(QuestionnaireFragment.kt:35)
        at com.google.android.fhir.datacapture.QuestionnaireFragment.access$getViewModel(QuestionnaireFragment.kt:34)
        at com.google.android.fhir.datacapture.QuestionnaireFragment$onViewCreated$3.invokeSuspend(QuestionnaireFragment.kt:71)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
        at androidx.lifecycle.DispatchQueue.drainQueue(DispatchQueue.kt:75)
        at androidx.lifecycle.DispatchQueue.resume(DispatchQueue.kt:54)
        at androidx.lifecycle.LifecycleController$observer$1.onStateChanged(LifecycleController.kt:40)
        at androidx.lifecycle.LifecycleRegistry$ObserverWithState.dispatchEvent(LifecycleRegistry.java:354)
        at androidx.lifecycle.LifecycleRegistry.forwardPass(LifecycleRegistry.java:265)
        at androidx.lifecycle.LifecycleRegistry.sync(LifecycleRegistry.java:307)
        at androidx.lifecycle.LifecycleRegistry.moveToState(LifecycleRegistry.java:148)
        at androidx.lifecycle.LifecycleRegistry.handleLifecycleEvent(LifecycleRegistry.java:134)
        at androidx.fragment.app.FragmentViewLifecycleOwner.handleLifecycleEvent(FragmentViewLifecycleOwner.java:88)
        at androidx.fragment.app.Fragment.restoreViewState(Fragment.java:653)
        at androidx.fragment.app.Fragment.restoreViewState(Fragment.java:3010)
        at androidx.fragment.app.Fragment.performActivityCreated(Fragment.java:3001)
        at androidx.fragment.app.FragmentStateManager.activityCreated(FragmentStateManager.java:580)
        at androidx.fragment.app.FragmentStateManager.moveToExpectedState(FragmentStateManager.java:285)
        at androidx.fragment.app.FragmentManager.executeOpsTogether(FragmentManager.java:2189)
        at androidx.fragment.app.FragmentManager.removeRedundantOperationsAndExecute(FragmentManager.java:2100)
        at androidx.fragment.app.FragmentManager.execPendingActions(FragmentManager.java:2002)
        at androidx.fragment.app.FragmentManager$5.run(FragmentManager.java:524)
        at android.os.Handler.handleCallback(Handler.java:938)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:223)
        at android.app.ActivityThread.main(ActivityThread.java:7656)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
     Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'boolean org.hl7.fhir.r4.model.Questionnaire$QuestionnaireItemType.equals(java.lang.Object)' on a null object reference
        at com.google.android.fhir.datacapture.QuestionnaireViewModel.validateQuestionnaireResponseItems(QuestionnaireViewModel.kt:320)
        at com.google.android.fhir.datacapture.QuestionnaireViewModel.<init>(QuestionnaireViewModel.kt:56)
        at java.lang.reflect.Constructor.newInstance0(Native Method)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:343)
        at androidx.lifecycle.SavedStateViewModelFactory.create(SavedStateViewModelFactory.java:122)
        at androidx.lifecycle.ViewModelProvider.get(ViewModelProvider.java:185) 
        at androidx.lifecycle.ViewModelProvider.get(ViewModelProvider.java:150) 
        at androidx.lifecycle.ViewModelLazy.getValue(ViewModelProvider.kt:54) 
        at androidx.lifecycle.ViewModelLazy.getValue(ViewModelProvider.kt:41) 
        at com.google.android.fhir.datacapture.QuestionnaireFragment.getViewModel(QuestionnaireFragment.kt:35) 
        at com.google.android.fhir.datacapture.QuestionnaireFragment.access$getViewModel(QuestionnaireFragment.kt:34) 
        at com.google.android.fhir.datacapture.QuestionnaireFragment$onViewCreated$3.invokeSuspend(QuestionnaireFragment.kt:71) 
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33) 
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106) 
        at androidx.lifecycle.DispatchQueue.drainQueue(DispatchQueue.kt:75) 
        at androidx.lifecycle.DispatchQueue.resume(DispatchQueue.kt:54) 
        at androidx.lifecycle.LifecycleController$observer$1.onStateChanged(LifecycleController.kt:40) 
        at androidx.lifecycle.LifecycleRegistry$ObserverWithState.dispatchEvent(LifecycleRegistry.java:354) 
        at androidx.lifecycle.LifecycleRegistry.forwardPass(LifecycleRegistry.java:265) 
        at androidx.lifecycle.LifecycleRegistry.sync(LifecycleRegistry.java:307) 
        at androidx.lifecycle.LifecycleRegistry.moveToState(LifecycleRegistry.java:148) 
        at androidx.lifecycle.LifecycleRegistry.handleLifecycleEvent(LifecycleRegistry.java:134) 
        at androidx.fragment.app.FragmentViewLifecycleOwner.handleLifecycleEvent(FragmentViewLifecycleOwner.java:88) 
        at androidx.fragment.app.Fragment.restoreViewState(Fragment.java:653) 
        at androidx.fragment.app.Fragment.restoreViewState(Fragment.java:3010) 
        at androidx.fragment.app.Fragment.performActivityCreated(Fragment.java:3001) 
        at androidx.fragment.app.FragmentStateManager.activityCreated(FragmentStateManager.java:580) 
        at androidx.fragment.app.FragmentStateManager.moveToExpectedState(FragmentStateManager.java:285) 
        at androidx.fragment.app.FragmentManager.executeOpsTogether(FragmentManager.java:2189) 
        at androidx.fragment.app.FragmentManager.removeRedundantOperationsAndExecute(FragmentManager.java:2100) 
        at androidx.fragment.app.FragmentManager.execPendingActions(FragmentManager.java:2002) 
        at androidx.fragment.app.FragmentManager$5.run(FragmentManager.java:524) 
        at android.os.Handler.handleCallback(Handler.java:938) 
        at android.os.Handler.dispatchMessage(Handler.java:99) 
        at android.os.Looper.loop(Looper.java:223) 
        at android.app.ActivityThread.main(ActivityThread.java:7656) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947) 
```

> @FikriMilano What happens if we set a type to prevent the NPE?

It will throw this error 
`Caused by: ca.uhn.fhir.parser.DataFormatException: [element="type"] Invalid attribute value "photo-capture": Unknown QuestionnaireItemType code 'photo-capture'`

--- a/android/engine/src/main/java/org/smartregister/fhircore/engine/ui/questionnaire/CustomPhotoCaptureFactory.kt
+++ b/android/engine/src/main/java/org/smartregister/fhircore/engine/ui/questionnaire/CustomPhotoCaptureFactory.kt
@@ -96,12 +96,17 @@ class CustomPhotoCaptureFactory(fragment: Fragment) :
         }
       val imageFileCompressed =
         compress(fragment.requireContext(), imageFile) { size(MAX_COMPRESSION_SIZE) }
-      questionnaireResponse.value =
-        Attachment().apply {
-          contentType = CONTENT_TYPE
-          data = imageFileCompressed.encodeToBase64()
-        }
+      val base64 = imageFileCompressed.encodeToBase64()
+      populateQuestionnaireResponse(base64)
     }
+  }
+
+  fun populateQuestionnaireResponse(base64: ByteArray) {
+    questionnaireResponse.value =
+      Attachment().apply {
+        contentType = CONTENT_TYPE
+        data = base64
+      }
   }
 
   override fun getQuestionnaireItemViewHolderDelegate(): QuestionnaireItemViewHolderDelegate =

--- a/android/engine/src/main/java/org/smartregister/fhircore/engine/ui/questionnaire/CustomPhotoCaptureFactory.kt
+++ b/android/engine/src/main/java/org/smartregister/fhircore/engine/ui/questionnaire/CustomPhotoCaptureFactory.kt
@@ -40,8 +40,8 @@ import id.zelory.compressor.saveBitmap
 import java.io.File
 import kotlinx.coroutines.launch
 import kotlinx.coroutines.withContext
+import org.hl7.fhir.r4.model.Attachment
 import org.hl7.fhir.r4.model.QuestionnaireResponse.QuestionnaireResponseItemAnswerComponent
-import org.hl7.fhir.r4.model.StringType
 import org.smartregister.fhircore.engine.R
 import org.smartregister.fhircore.engine.util.DefaultDispatcherProvider
 import org.smartregister.fhircore.engine.util.DispatcherProvider
@@ -96,7 +96,11 @@ class CustomPhotoCaptureFactory(fragment: Fragment) :
         }
       val imageFileCompressed =
         compress(fragment.requireContext(), imageFile) { size(MAX_COMPRESSION_SIZE) }
-      questionnaireResponse.value = StringType(imageFileCompressed.encodeToBase64())
+      questionnaireResponse.value =
+        Attachment().apply {
+          contentType = CONTENT_TYPE
+          data = imageFileCompressed.encodeToBase64()
+        }
     }
   }
 
@@ -140,5 +144,6 @@ class CustomPhotoCaptureFactory(fragment: Fragment) :
     private const val EXTRA_IMAGE = "data"
     private const val MAX_COMPRESSION_SIZE: Long = 64_000 // 64 KB
     private const val PREFIX_BITMAP = "BITMAP_"
+    private const val CONTENT_TYPE = "image/jpg"
   }
 }

--- a/android/engine/src/main/java/org/smartregister/fhircore/engine/ui/questionnaire/FhirCoreQuestionnaireFragment.kt
+++ b/android/engine/src/main/java/org/smartregister/fhircore/engine/ui/questionnaire/FhirCoreQuestionnaireFragment.kt
@@ -19,34 +19,25 @@ package org.smartregister.fhircore.engine.ui.questionnaire
 import com.google.android.fhir.datacapture.QuestionnaireFragment
 import com.google.android.fhir.datacapture.common.datatype.asStringValue
 import com.google.android.fhir.datacapture.contrib.QuestionnaireItemBarCodeReaderViewHolderFactory
-import com.google.android.fhir.datacapture.views.QuestionnaireItemViewHolderFactory
 
 class FhirCoreQuestionnaireFragment : QuestionnaireFragment() {
 
   override fun getCustomQuestionnaireItemViewHolderFactoryMatchers():
     List<QuestionnaireItemViewHolderFactoryMatcher> {
     return listOf(
-      addQuestionnaireItemViewHolderFactoryMatcher(
-        QuestionnaireItemBarCodeReaderViewHolderFactory,
-        BARCODE_URL,
-        BARCODE_NAME
-      ),
-      addQuestionnaireItemViewHolderFactoryMatcher(
-        CustomPhotoCaptureFactory(this),
-        PHOTO_CAPTURE_NAME,
-        PHOTO_CAPTURE_URL
-      ),
+      QuestionnaireItemViewHolderFactoryMatcher(QuestionnaireItemBarCodeReaderViewHolderFactory) {
+        questionnaireItem ->
+        questionnaireItem.getExtensionByUrl(BARCODE_URL).let {
+          if (it == null) false else it.value.asStringValue() == BARCODE_NAME
+        }
+      },
+      QuestionnaireItemViewHolderFactoryMatcher(CustomPhotoCaptureFactory(this)) { questionnaireItem
+        ->
+        questionnaireItem.getExtensionByUrl(PHOTO_CAPTURE_URL).let {
+          if (it == null) false else it.value.asStringValue() == PHOTO_CAPTURE_NAME
+        }
+      }
     )
-  }
-
-  private fun addQuestionnaireItemViewHolderFactoryMatcher(
-    factory: QuestionnaireItemViewHolderFactory,
-    url: String,
-    name: String
-  ): QuestionnaireItemViewHolderFactoryMatcher {
-    return QuestionnaireItemViewHolderFactoryMatcher(factory) { questionnaireItem ->
-      questionnaireItem.getExtensionByUrl(url)?.let { it.value.asStringValue() == name } ?: false
-    }
   }
 
   companion object {

--- a/android/engine/src/main/java/org/smartregister/fhircore/engine/util/extension/FileExtension.kt
+++ b/android/engine/src/main/java/org/smartregister/fhircore/engine/util/extension/FileExtension.kt
@@ -22,12 +22,11 @@ import java.io.ByteArrayOutputStream
 import java.io.File
 import java.io.FileInputStream
 
-fun File.encodeToBase64(): String {
-  return ByteArrayOutputStream()
-    .use { byteStream ->
-      Base64OutputStream(byteStream, Base64.DEFAULT).use { base64Stream ->
-        FileInputStream(this).copyTo(base64Stream)
-      }
+fun File.encodeToBase64(): ByteArray {
+  return ByteArrayOutputStream().use { byteStream ->
+    Base64OutputStream(byteStream, Base64.DEFAULT).use { base64Stream ->
+      FileInputStream(this).copyTo(base64Stream)
     }
-    .toString()
+    return@use byteStream.toByteArray()
+  }
 }

--- a/android/engine/src/test/java/org/smartregister/fhircore/engine/ui/questionnaire/CustomPhotoCaptureFactoryTest.kt
+++ b/android/engine/src/test/java/org/smartregister/fhircore/engine/ui/questionnaire/CustomPhotoCaptureFactoryTest.kt
@@ -31,9 +31,9 @@ import io.mockk.mockk
 import io.mockk.slot
 import io.mockk.spyk
 import io.mockk.verify
+import org.hl7.fhir.r4.model.Attachment
 import org.hl7.fhir.r4.model.Questionnaire
 import org.hl7.fhir.r4.model.QuestionnaireResponse
-import org.hl7.fhir.r4.model.StringType
 import org.junit.Assert.assertEquals
 import org.junit.Test
 import org.smartregister.fhircore.engine.robolectric.RobolectricTest
@@ -66,17 +66,23 @@ class CustomPhotoCaptureFactoryTest : RobolectricTest() {
   }
 
   @Test
-  fun testQuestionnaireResponseShouldReturnStringTypeValue() {
+  fun testQuestionnaireResponseShouldReturnAttachmentTypeValue() {
     val fragment = FhirCoreQuestionnaireFragment()
     val photoCaptureFactory = spyk(CustomPhotoCaptureFactory(fragment))
+    val file = "file".toByteArray()
 
     every { photoCaptureFactory.questionnaireResponse } returns
       QuestionnaireResponse.QuestionnaireResponseItemAnswerComponent().apply {
-        value = StringType("file")
+        value =
+          Attachment().apply {
+            contentType = "image/jpg"
+            data = file
+          }
       }
 
     val response = photoCaptureFactory.questionnaireResponse
-    assertEquals("file", response.valueStringType.value)
+    assertEquals("image/jpg", response.valueAttachment.contentType)
+    assertEquals(file, response.valueAttachment.data)
   }
 
   @Test

--- a/android/engine/src/test/java/org/smartregister/fhircore/engine/ui/questionnaire/CustomPhotoCaptureFactoryTest.kt
+++ b/android/engine/src/test/java/org/smartregister/fhircore/engine/ui/questionnaire/CustomPhotoCaptureFactoryTest.kt
@@ -66,6 +66,19 @@ class CustomPhotoCaptureFactoryTest : RobolectricTest() {
   }
 
   @Test
+  fun testPopulateQuestionnaireResponseShouldMatch() {
+    val fragment = spyk<FhirCoreQuestionnaireFragment>()
+    val photoCaptureFactory = spyk(CustomPhotoCaptureFactory(fragment))
+    val base64 = "file".toByteArray()
+
+    photoCaptureFactory.populateQuestionnaireResponse(base64)
+
+    assertEquals(photoCaptureFactory.questionnaireResponse.valueAttachment.data, base64)
+
+    verify { photoCaptureFactory.populateQuestionnaireResponse(any()) }
+  }
+
+  @Test
   fun testQuestionnaireResponseShouldReturnAttachmentTypeValue() {
     val fragment = FhirCoreQuestionnaireFragment()
     val photoCaptureFactory = spyk(CustomPhotoCaptureFactory(fragment))

--- a/android/engine/src/test/java/org/smartregister/fhircore/engine/ui/questionnaire/QuestionnaireViewModelTest.kt
+++ b/android/engine/src/test/java/org/smartregister/fhircore/engine/ui/questionnaire/QuestionnaireViewModelTest.kt
@@ -355,8 +355,8 @@ class QuestionnaireViewModelTest : RobolectricTest() {
       questionnaireResponse
     )
 
-    coVerify { defaultRepo.addOrUpdate(questionnaireResponse) }
     coVerify { defaultRepo.addOrUpdate(patient) }
+    coVerify { defaultRepo.addOrUpdate(questionnaireResponse) }
     coVerify(timeout = 2000) { ResourceMapper.extract(any(), any(), any(), any()) }
 
     unmockkObject(ResourceMapper)

--- a/android/engine/src/test/java/org/smartregister/fhircore/engine/util/extension/FileExtensionTest.kt
+++ b/android/engine/src/test/java/org/smartregister/fhircore/engine/util/extension/FileExtensionTest.kt
@@ -16,11 +16,10 @@
 
 package org.smartregister.fhircore.engine.util.extension
 
-import io.mockk.every
 import io.mockk.mockkStatic
 import io.mockk.verify
 import java.io.File
-import org.junit.Assert.assertEquals
+import java.io.FileOutputStream
 import org.junit.Test
 
 class FileExtensionTest {
@@ -29,14 +28,14 @@ class FileExtensionTest {
   fun testFileShouldReturnBase64Encoded() {
     mockkStatic(File::encodeToBase64)
 
-    val file = "Zmls".toByteArray()
-    val filePath = javaClass.getResource("/sample/file.txt").path
+    val file = File.createTempFile("sample_file", ".txt")
+    val writer = FileOutputStream(file)
+    writer.write("file".toByteArray())
 
-    every { File(filePath).encodeToBase64() } returns file
+    file.encodeToBase64()
 
-    val base64 = File(filePath).encodeToBase64()
-    assertEquals(file, base64)
+    verify { file.encodeToBase64() }
 
-    verify { File(filePath).encodeToBase64() }
+    file.delete()
   }
 }

--- a/android/engine/src/test/java/org/smartregister/fhircore/engine/util/extension/FileExtensionTest.kt
+++ b/android/engine/src/test/java/org/smartregister/fhircore/engine/util/extension/FileExtensionTest.kt
@@ -29,11 +29,13 @@ class FileExtensionTest {
   fun testFileShouldReturnBase64Encoded() {
     mockkStatic(File::encodeToBase64)
 
+    val file = "Zmls".toByteArray()
     val filePath = javaClass.getResource("/sample/file.txt").path
-    every { File(filePath).encodeToBase64() } returns "Zmls"
+
+    every { File(filePath).encodeToBase64() } returns file
 
     val base64 = File(filePath).encodeToBase64()
-    assertEquals("Zmls", base64)
+    assertEquals(file, base64)
 
     verify { File(filePath).encodeToBase64() }
   }

